### PR TITLE
DependenciesScanner: report command-line arguments for building pcms explicitly

### PR DESCRIPTION
**Pull request by nkcsgexi**
Clang PCM dependencies are currently implicitly built by the compiler. To allow build-system level to build these PCMs explicitly, we need to report the command line arguments from the compiler to the build system via a JSON file emitted from the dependencies scanner.

This PR combines and emits two sets of Clang compiler arguments: (1) arguments Swift used to construct ClangImporter instance, and (2) arguments collected from running Clang's fast dependency scanner. A high level build system can explicitly schedule PCM building by feeding these arguments to a Swift front-end invocation with an action kind `-emit-pcm`.

For testing purposes, a Swift script is also added to build modules explicitly by using these arguments from the JSON file.

rdar://62612967
rdar://62613017

### CODE REVIEW

File: lib/ClangImporter/ClangImporter.cpp [hunk 1]
```diff
@@ -927,6 +922,49 @@ ClangImporter::create(ASTContext &ctx, const ClangImporterOptions &importerOpts,
...
   }
   addCommonInvocationArguments(invocationArgStrs, ctx, importerOpts);
+  return invocationArgStrs;
+}
+
+std::unique_ptr<clang::CompilerInvocation>
+ClangImporter::createClangInvocation(ClangImporter *importer,
+                                     const ClangImporterOptions &importerOpts,
+                                     ArrayRef<std::string> invocationArgStrs,
+                                     std::vector<std::string> &CC1Args) {
```
Comment (Bigcheese) [inline lib/ClangImporter/ClangImporter.cpp:932]: You should probably make this optional as you have clients that are uninterested in it.

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 2]
```diff
@@ -204,14 +204,29 @@ static void recordModuleDependencies(
     for (const auto &fileDep : clangModuleDep.FileDeps) {
       fileDeps.push_back(fileDep.getKey().str());
     }
-
+    std::vector<std::string> allArgs = Impl.ClangArgs;
+    for(auto arg: clangModuleDep.NonPathCommandLine) {
+      allArgs.push_back("-Xclang");
+      allArgs.push_back(arg);
+    }
```
Comment (Bigcheese) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp]: The arguments in `NonPathCommandLine` are intended to be directly appended to a `-cc1` command line, not to the driver.

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 3]
```diff
@@ -204,14 +204,29 @@ static void recordModuleDependencies(
...
+    ClangImporterOptions Opts;
+    std::vector<std::string> cc1Args;
+    createClangInvocation(this, Opts, allArgs, cc1Args);
+    std::vector<std::string> swiftArgs;
+    swiftArgs.reserve(cc1Args.size() * 3);
+    for (auto arg: cc1Args) {
+      swiftArgs.push_back("-Xcc");
+      swiftArgs.push_back("-Xclang");
```
Comment (Bigcheese) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp]: I'm not sure how `-Xcc` works in the Swift driver, but given this is a `-cc1` command line you can't use `-Xclang`, as that's for forwarding driver options to the frontend.

**Review by Bigcheese — CHANGES_REQUESTED** [review]
This also needs to handle the other stuff that `getFullCommandLine` does for `-fmodule-map-file=` and `-fmodule-file=`.

**Comment by Bigcheese** [comment]
> This also needs to handle the other stuff that `getFullCommandLine` does for `-fmodule-map-file=` and `-fmodule-file=`.

Actually, I guess this would be handled at a higher level.

**Comment by nkcsgexi** [comment]
Yeah, I think `-fmodule-map-file=` and `-fmodule-file=` are directly handled by Swift in `emitPrecompiledModule` function.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ClangImporter.cpp (hunk 1 above)
Comment (nkcsgexi) [reply lib/ClangImporter/ClangImporter.cpp]: That's a good idea. Fixed.

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test OS X platform

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test OS X  platform

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test OS X platform

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
> rdar://62612967
> rdar://62613017

Please replace this with a full description of which this pull request does. The radar numbers are good for automation, and can be at the end, but the pull request description should be understandable to anyone who cannot follow that link.

**Comment by nkcsgexi** [comment]
> > rdar://62612967
> > rdar://62613017
> 
> Please replace this with a full description of which this pull request does. The radar numbers are good for automation, and can be at the end, but the pull request description should be understandable to anyone who cannot follow that link.

Good point. I've updated the PR description.

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Pull request merged**